cabal-install/src/Distribution/Client/InLibrary.hs

```diff
@@ -0,0 +1,352 @@
+import qualified Distribution.Simple.Test as Cabal
+import Distribution.Simple.Utils
+import Distribution.System (Platform)
+import Distribution.Types.BuildType
+import Distribution.Types.ComponentRequestedSpec
+import qualified Distribution.Types.LocalBuildConfig as LBC
+import Distribution.Types.LocalBuildInfo
+import Distribution.Utils.Path
+  ( relativeSymbolicPath
+  )
+
+import qualified Data.Set as Set
+
+--------------------------------------------------------------------------------
+-- Configure
+
+data LibraryConfigureInputs = LibraryConfigureInputs
+  { compiler :: Compiler
+  , platform :: Platform
+  , buildType :: BuildType
+  , compRequested :: Maybe PD.ComponentName
+  , localBuildConfig :: LBC.LocalBuildConfig
+  , packageDBStack :: PackageDBStack
+  , packageDescription :: PD.PackageDescription
+  , gPackageDescription :: PD.GenericPackageDescription
+  , flagAssignment :: PD.FlagAssignment
+  }
+
+libraryConfigureInputsFromElabPackage
+  :: ProgramDb
+  -> ElaboratedSharedConfig
+  -> ElaboratedReadyPackage
+  -> [String]
+  -- ^ targets
+  -> LibraryConfigureInputs
+libraryConfigureInputsFromElabPackage
+  progDb
+  -- NB: don't use the ProgramDb from the ElaboratedSharedConfig;
+  -- that one is only for the compiler itself and not for the package.
+  ElaboratedSharedConfig
+    { pkgConfigPlatform = platform
+    , pkgConfigCompiler = compiler
+    }
+  (ReadyPackage pkg)
+  userTargets =
+    LibraryConfigureInputs
+      { compiler
+      , platform
+      , buildType = PD.buildType pkgDescr
+      , compRequested =
+          case elabPkgOrComp pkg of
+            ElabComponent elabComp
+              | Just elabCompNm <- compComponentName elabComp ->
+                  Just elabCompNm
+            _ -> Nothing
+      , localBuildConfig =
+          LBC.LocalBuildConfig
+            { LBC.extraConfigArgs = userTargets
+            , LBC.withPrograms = progDb
+            , LBC.withBuildOptions = elabBuildOptions pkg
+            }
+      , packageDBStack = elabBuildPackageDBStack pkg
+      , packageDescription = pkgDescr
+      , gPackageDescription = gpkgDescr
+      , flagAssignment = elabFlagAssignment pkg
+      }
+    where
+      pkgDescr = elabPkgDescription pkg
+      gpkgDescr = elabGPkgDescription pkg
+
+configure
+  :: LibraryConfigureInputs
+  -> Cabal.ConfigFlags
+  -> IO LocalBuildInfo
+configure
+  LibraryConfigureInputs
+    { platform
+    , compiler
+    , buildType = bt
+    , compRequested = mbComp
+    , localBuildConfig = lbc0
+    , packageDBStack = packageDBs
+    , packageDescription = pkgDesc
+    , gPackageDescription = gpkgDescr
+    , flagAssignment
+    }
+  cfg =
+    -- SetupHooks TODO: the following code should not live in cabal-install.
+    -- We should be able to directly call into the library,
+    -- similar to what we do for other phases (see e.g. inLibraryBuild).
+    --
+    -- The issue is mainly about 'finalizeAndConfigurePackage' vs 'configurePackage'.
```

Comment:
Should this be addressed?

Reply:
In theory yes, but I don't know how, because we have different information in `cabal-install` so we can't easily just call the exact same functions. The functions we call in `Cabal` would need to be restructured a bit to make this easier, but I'm not sure how best to do that. I'm planning to leave this TODO in the code.